### PR TITLE
(feat) Restrict attachments to configured allowed extensions

### PR DESCRIPTION
**Pull request by icrc-loliveira**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [x] My work includes tests or is validated by existing tests.

## Summary
This PR created a new configuration 'allowedExtensions' that allow filter by the extensions that can be attached to openMRS.

## Screenshots
image
image

### CODE REVIEW

**Comment by icrc-loliveira** [comment]
@ibacher

File: packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  value: string;
+  description: string;
+}
+
+/**
+ * React hook that takes returns the allowed file extensions
+ * @returns String array containing the `allowedExtensions`, `isLoading` loading status, `error`
+ */
+export const useAllowedExtensions = () => {
+  const customRepresentation = 'custom:(uuid,property,value,description)';
```
Comment (ibacher) [inline packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts:17]: Do we need the description?

File: packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  const { data, error, isLoading } = useSWRImmutable<{ data: { results: Array<GlobalProperty> } }>(
+    `/ws/rest/v1/systemsetting?&v=${customRepresentation}&q=${allowedExtensionsGlobalProperty}`,
+    openmrsFetch,
+  );
+
+  const extensions: Array<string> =
+    data?.data?.results?.length > 0
+      ? data?.data?.results[0].value?.split(',').map<string>((ext) => (ext = '.' + ext))
```
Comment (ibacher) [inline packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts:25]:
```suggestion
      ? data?.data?.results[0].value?.split(',').map((ext) => '.' + ext)
```

File: packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * @returns String array containing the `allowedExtensions`, `isLoading` loading status, `error`
+ */
+export const useAllowedExtensions = () => {
+  const customRepresentation = 'custom:(uuid,property,value,description)';
+  const { data, error, isLoading } = useSWRImmutable<{ data: { results: Array<GlobalProperty> } }>(
+    `/ws/rest/v1/systemsetting?&v=${customRepresentation}&q=${allowedExtensionsGlobalProperty}`,
+    openmrsFetch,
+  );
+
+  const extensions: Array<string> =
```
Comment (ibacher) [inline packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts:23]: Why do we have this if it's never returned or used?

File: packages/esm-patient-attachments-app/src/camera-media-uploader/media-uploader.component.tsx [hunk 4]
```diff
@@ -50,7 +64,7 @@ const MediaUploaderComponent = () => {
       </p>
       <div className={styles.uploadFile}>
         <FileUploaderDropContainer
-          accept={allowedExtensions ?? ['*']}
+          accept={allowedExtensions?.map<string>((ext) => (ext = '.' + ext)) ?? ['*']}
```
Comment (ibacher) [inline packages/esm-patient-attachments-app/src/camera-media-uploader/media-uploader.component.tsx:67]:
```suggestion
          accept={allowedExtensions?.map((ext) =>  '.' + ext) || ['*']}
```

File: packages/esm-patient-attachments-app/src/constants.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export const allowedExtensionsGlobalProperty: string = 'attachments.allowedFileExtensions';
```
Comment (ibacher) [inline packages/esm-patient-attachments-app/src/constants.ts:1]: I don't think we need a constants file unless this is used in multiple places. Just keep this local to the hook that uses it.

**Review by ibacher — CHANGES_REQUESTED** [review]
Thanks @icrc-loliveira!

File: packages/esm-patient-attachments-app/src/attachments/attachments-overview.component.tsx [hunk 6]
```diff
@@ -19,6 +19,8 @@ import AttachmentsGridOverview from './attachments-grid-overview.component';
 import AttachmentsTableOverview from './attachments-table-overview.component';
 import AttachmentPreview from './image-preview.component';
 import styles from './attachments-overview.scss';
+import { List, Thumbnail_2, Add } from '@carbon/react/icons';
```
Comment (ibacher) [inline packages/esm-patient-attachments-app/src/attachments/attachments-overview.component.tsx:22]: This should still be up around line 3

File: packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * Global property created by the openmrs-module-attachment
+ */
+const allowedExtensionsGlobalProperty: string = 'attachments.allowedFileExtensions';
+
+/**
+ * React hook that takes returns the allowed file extensions
+ * @returns String array containing the `allowedExtensions`, `isLoading` loading status, `error`
+ */
+export const useAllowedExtensions = () => {
+  const customRepresentation = 'custom:(uuid,property,value)';
```
Comment (ibacher) [inline packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts:20]:
```suggestion
  const customRepresentation = 'custom:(value)';
```

Comment (ibacher) [reply packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts:20]: We only really need the value right?

File: packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  const { data, error, isLoading } = useSWRImmutable<{ data: { results: Array<GlobalProperty> } }>(
+    `/ws/rest/v1/systemsetting?&v=${customRepresentation}&q=${allowedExtensionsGlobalProperty}`,
+    openmrsFetch,
+  );
+
+  return {
+    isLoading,
+    allowedExtensions: data?.data?.results?.length > 0 ? data?.data?.results[0].value?.split(',') : undefined,
```
Comment (ibacher) [inline packages/esm-patient-attachments-app/src/attachments/use-allowed-extensions.ts:28]: Deal with an empty string:

```suggestion
    allowedExtensions: data?.data?.results?.length > 0 ? data?.data?.results[0].value?.split(',') || undefined : undefined,
```

**Review by ibacher — CHANGES_REQUESTED** [review]
Couple of small things. Has this been tested in a system without that global property set?

**Review by ibacher — APPROVED** [review]

**Pull request merged**